### PR TITLE
Add embedding_bag meta functions

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79026
* __->__ #78997
* #78996
* #78993
* #78992

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
UGH what a function.

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -331,3 +331,71 @@ def meta_cdist_forward(x1, x2, p, compute_mode):
     output_shape = list(torch.broadcast_shapes(batch_tensor1, batch_tensor2))
     output_shape.extend([r1, r2])
     return x1.new_empty(output_shape)
+
+@torch.library.impl(meta_lib, "_embedding_bag")
+def meta_embedding_bag(
+    weight, indices, offsets, scale_grad_by_freq=False, mode=0,
+    sparse=False, per_sample_weights=None, include_last_offset=False, padding_idx=-1
+):
+    check(indices.dtype in (torch.long, torch.int), lambda: f"expected indices to be long or int, got {indices.dtype}")
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_meta_registrations.py:340]: The alternative is I can port the TensorArg error reporting framework into Python. WDYT.

[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_meta_registrations.py:340]: Is it used a lot? If it is, I'd say go for it. I, personally, have never used it, but idk

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_meta_registrations.py:340]: It's used inconsistently. I don't think it's the right abstraction, but there are a least a dozen or more uses of it

[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_meta_registrations.py:340]: I'd say we create our own auxiliary functions as we go but with the same messages as the previous ones?

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_meta_registrations.py:340]: The problem is I don't really want to make a complicated auxiliary function without frying the general problem of how we should structure them, but if we want to use the exactly same messages it's a lot of repeated typing lol. Also the C++ error messages need to make up for lack of Python backtrace but it is less important here.

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -331,3 +331,71 @@ def meta_cdist_forward(x1, x2, p, compute_mode):
...
+    )
+
+    num_bags = offsets.size(0)
+    if include_last_offset:
+        check(num_bags >= 1, lambda: "include_last_offset: numBags should be at least 1")
+        num_bags -= 1
+
+    output = weight.new_empty(num_bags, weight.size(1))
+    MODE_SUM, MODE_MEAN, MODE_MAX = range(3)
+
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_meta_registrations.py:354]: This is still missing I believe?

**Review by Chillee — APPROVED** [review]

**Review by lezcano — APPROVED** [review]
cool

**Pull request closed without merging**